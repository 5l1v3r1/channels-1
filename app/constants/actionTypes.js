// @flow
const actionTypes = {
  FETCH_ACCESS_TOKEN_REQUEST: 'FETCH_ACCESS_TOKEN_REQUEST',
  FETCH_ACCESS_TOKEN_SUCCESS: 'FETCH_ACCESS_TOKEN_SUCCESS',
  FETCH_ACCESS_TOKEN_FAILURE: 'FETCH_ACCESS_TOKEN_FAILURE',
  FETCH_CHANNELS_REQUEST: 'FETCH_CHANNELS_REQUEST',
  FETCH_CHANNELS_SUCCESS: 'FETCH_CHANNELS_SUCCESS',
  FETCH_CHANNELS_FAILURE: 'FETCH_CHANNELS_FAILURE',
  UPDATE_ACTIVE_CHANNEL: 'UPDATE_ACTIVE_CHANNEL'
};

export default actionTypes;
