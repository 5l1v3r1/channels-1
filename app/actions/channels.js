// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { token } = getState().accessToken;
    const subscriptions = fetchSubscriptions(token);
    return subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(token, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          return dispatch(receiveChannels(json.items));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

export function fetchSubscriptions(token: string): Promise<any> {
  const uri = getSubscriptionsUri(token);
  return apiRequest(uri, Methods.GET);
}

export function requestChannels(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_REQUEST
  };
}

export function receiveChannels(items: any): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_SUCCESS,
    payload: items
  };
}

export function channelsError(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_FAILURE
  };
}

export function updateActiveChannel(channelId: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { activeId } = getState().channels;
    if (activeId !== channelId) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_CHANNEL,
        payload: channelId
      });
    }
  };
}
