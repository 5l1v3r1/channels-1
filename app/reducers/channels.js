// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action, ChannelsState } from '../constants/typeAliases';
import ChannelType from '../dataTypes/channelType';
import { stringOrEmpty } from '../utils/utils';

function byId(state: { [string]: ChannelType } = {}, { type, payload }: Action): { [string]: ChannelType } {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_SUCCESS:
      const nextState = { ...state };
      payload.forEach(item => {
        nextState[item.id] = ChannelType.from(item);
      });
      return nextState;
    default:
      return state;
  }
}

function allIds(state: string[] = [], { type, payload }: Action): string[] {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_SUCCESS:
      return payload.map(item => stringOrEmpty(item.id));
    default:
      return state;
  }
}

function activeId(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.UPDATE_ACTIVE_CHANNEL:
      return payload;
    default:
      return state;
  }
}

function isFetching(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_REQUEST:
      return true;
    case actionTypes.FETCH_CHANNELS_SUCCESS:
    case actionTypes.FETCH_CHANNELS_FAILURE:
      return false;
    default:
      return state;
  }
}

function error(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_CHANNELS_REQUEST:
    case actionTypes.FETCH_CHANNELS_SUCCESS:
      return false;
    case actionTypes.FETCH_CHANNELS_FAILURE:
      return true;
    default:
      return state;
  }
}

const channels = combineReducers({
  byId,
  allIds,
  activeId,
  isFetching,
  error
});

export function getChannels(state: ChannelsState) {
  return state.allIds.map(id => state.byId[id]);
}

export function getActiveChannel(state: ChannelsState) {
  return state.byId[state.activeId];
}

export default channels;
