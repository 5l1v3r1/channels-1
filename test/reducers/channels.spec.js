import channels from '../../app/reducers/channels';
import actionTypes from '../../app/constants/actionTypes';
import ChannelType from '../../app/dataTypes/channelType';

describe('channels reducer', () => {
  it('should return the initial state', () => {
    expect(channels(undefined, {})).toEqual({
      byId: {},
      allIds: [],
      activeId: '',
      isFetching: false,
      error: false
    });
  });

  it('should handle FETCH_CHANNELS_REQUEST', () => {
    const action = { type: actionTypes.FETCH_CHANNELS_REQUEST };
    expect(channels(undefined, action)).toEqual({
      byId: {},
      allIds: [],
      activeId: '',
      isFetching: true,
      error: false
    });
  });

  it('should handle FETCH_CHANNELS_SUCCESS', () => {
    const state = {
      byId: {},
      allIds: [],
      activeId: '',
      isFetching: true,
      error: false
    };
    const action = {
      type: actionTypes.FETCH_CHANNELS_SUCCESS,
      payload: [
        { id: 'UCyIe-61Y8C4_o-zZCtO4ETQ' },
        { id: 'UCO1cgjhGzsSYb1rsB4bFe4Q' }
      ]
    };

    ChannelType.from = jest.fn(item => item);

    expect(channels(state, action)).toEqual({
      byId: {
        'UCyIe-61Y8C4_o-zZCtO4ETQ': {
          id: 'UCyIe-61Y8C4_o-zZCtO4ETQ'
        },
        'UCO1cgjhGzsSYb1rsB4bFe4Q': {
          id: 'UCO1cgjhGzsSYb1rsB4bFe4Q'
        }
      },
      allIds: [
        'UCyIe-61Y8C4_o-zZCtO4ETQ',
        'UCO1cgjhGzsSYb1rsB4bFe4Q'
      ],
      activeId: '',
      isFetching: false,
      error: false
    });
  });

  it('should handle FETCH_CHANNELS_FAILURE', () => {
    const state = {
      byId: {},
      allIds: [],
      activeId: '',
      isFetching: true,
      error: false
    };
    const action = { type: actionTypes.FETCH_CHANNELS_FAILURE };
    expect(channels(state, action)).toEqual({
      byId: {},
      allIds: [],
      activeId: '',
      isFetching: false,
      error: true
    });
  });

  it('should handle UPDATE_ACTIVE_CHANNEL', () => {
    const state = {
      byId: {
        'UCyIe-61Y8C4_o-zZCtO4ETQ': {
          id: 'UCyIe-61Y8C4_o-zZCtO4ETQ'
        },
        'UCO1cgjhGzsSYb1rsB4bFe4Q': {
          id: 'UCO1cgjhGzsSYb1rsB4bFe4Q'
        }
      },
      allIds: [
        'UCyIe-61Y8C4_o-zZCtO4ETQ',
        'UCO1cgjhGzsSYb1rsB4bFe4Q'
      ],
      activeId: 'UCyIe-61Y8C4_o-zZCtO4ETQ',
      isFetching: false,
      error: false
    };
    const action = {
      type: actionTypes.UPDATE_ACTIVE_CHANNEL,
      payload: 'UCO1cgjhGzsSYb1rsB4bFe4Q'
    };
    expect(channels(state, action)).toEqual({
      byId: {
        'UCyIe-61Y8C4_o-zZCtO4ETQ': {
          id: 'UCyIe-61Y8C4_o-zZCtO4ETQ'
        },
        'UCO1cgjhGzsSYb1rsB4bFe4Q': {
          id: 'UCO1cgjhGzsSYb1rsB4bFe4Q'
        }
      },
      allIds: [
        'UCyIe-61Y8C4_o-zZCtO4ETQ',
        'UCO1cgjhGzsSYb1rsB4bFe4Q'
      ],
      activeId: 'UCO1cgjhGzsSYb1rsB4bFe4Q',
      isFetching: false,
      error: false
    });
  });

  it('should handle unknown action type', () => {
    const action = { type: 'unknown' };
    expect(channels(undefined, action)).toEqual({
      byId: {},
      allIds: [],
      activeId: '',
      isFetching: false,
      error: false
    });
  });
});
