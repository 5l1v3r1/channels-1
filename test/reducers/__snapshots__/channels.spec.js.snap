// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channels reducer should handle FETCH_CHANNELS_FAILURE 1`] = `
Object {
  "activeId": "",
  "allIds": Array [],
  "byId": Object {},
  "error": true,
  "isFetching": false,
}
`;

exports[`channels reducer should handle FETCH_CHANNELS_REQUEST 1`] = `
Object {
  "activeId": "",
  "allIds": Array [],
  "byId": Object {},
  "error": false,
  "isFetching": true,
}
`;

exports[`channels reducer should handle FETCH_CHANNELS_SUCCESS 1`] = `
Object {
  "activeId": "",
  "allIds": Array [
    "UCyIe-61Y8C4_o-zZCtO4ETQ",
    "UCO1cgjhGzsSYb1rsB4bFe4Q",
  ],
  "byId": Object {
    "UCO1cgjhGzsSYb1rsB4bFe4Q": Object {
      "id": "UCO1cgjhGzsSYb1rsB4bFe4Q",
    },
    "UCyIe-61Y8C4_o-zZCtO4ETQ": Object {
      "id": "UCyIe-61Y8C4_o-zZCtO4ETQ",
    },
  },
  "error": false,
  "isFetching": false,
}
`;

exports[`channels reducer should handle UPDATE_ACTIVE_CHANNEL 1`] = `
Object {
  "activeId": "UCO1cgjhGzsSYb1rsB4bFe4Q",
  "allIds": Array [
    "UCyIe-61Y8C4_o-zZCtO4ETQ",
    "UCO1cgjhGzsSYb1rsB4bFe4Q",
  ],
  "byId": Object {
    "UCO1cgjhGzsSYb1rsB4bFe4Q": Object {
      "id": "UCO1cgjhGzsSYb1rsB4bFe4Q",
    },
    "UCyIe-61Y8C4_o-zZCtO4ETQ": Object {
      "id": "UCyIe-61Y8C4_o-zZCtO4ETQ",
    },
  },
  "error": false,
  "isFetching": false,
}
`;

exports[`channels reducer should handle unknown action type 1`] = `
Object {
  "activeId": "",
  "allIds": Array [],
  "byId": Object {},
  "error": false,
  "isFetching": false,
}
`;

exports[`channels reducer should return the initial state 1`] = `
Object {
  "activeId": "",
  "allIds": Array [],
  "byId": Object {},
  "error": false,
  "isFetching": false,
}
`;
